src/app/public/modules/progress-indicator/progress-indicator-nav-button/progress-indicator-nav-button.component.ts

```diff
@@ -1,116 +1,233 @@
+
+  @Input()
+  public progressIndicator: SkyProgressIndicatorComponent;
+
+  public get cssClassNames(): string {
+    const buttonType = this.buttonType;
+
+    const classNames = [
+      `sky-progress-indicator-nav-button-${this.buttonType}`
+    ];
+
+    switch (buttonType) {
+      case 'next':
+      case 'finish':
+        classNames.push('sky-btn-primary');
+        break;
+
+      case 'reset':
+        classNames.push('sky-btn-link');
+        break;
+
+      default:
+        classNames.push('sky-btn-default');
+        break;
+    }
+
+    return classNames.join(' ');
+  }
+
+  public get buttonLabelResourceString(): string {
+    return `skyux_progress_indicator_navigator_${this.buttonType}`;
+  }
+
+  public set isVisible(value: boolean) {
+    this._isVisible = value;
+    this.changeDetector.markForCheck();
+  }
+
+  public get isVisible(): boolean {
+    return this._isVisible || false;
   }
 
-  public buttonTextStream = new BehaviorSubject<string>('');
+  private lastProgressChange: SkyProgressIndicatorChange;
+  private ngUnsubscribe = new Subject<void>();
 
-  private idle = new Subject();
-  private _buttonType: string;
+  private _buttonType: SkyProgressIndicatorNavButtonType;
   private _disabled: boolean;
-  private activeIndex = 0;
+  private _isVisible: boolean;
 
   constructor(
-    private resourcesService: SkyLibResourcesService,
-    private changeDetector: ChangeDetectorRef
+    private changeDetector: ChangeDetectorRef,
+    @Optional() private parentComponent: SkyProgressIndicatorComponent
   ) { }
 
   public ngOnInit(): void {
+    if (!this.progressIndicator) {
+      if (!this.parentComponent) {
+        throw new Error(
+          'The `<sky-progress-indicator-nav-button>` component requires a reference to ' +
+          'the `<sky-progress-indicator>` component it controls. For example:\n' +
+          '<sky-progress-indicator\n' +
+          '  #myProgressIndicator\n' +
+          '>\n' +
+          '</sky-progress-indicator>\n' +
+          '<sky-progress-indicator-nav-button\n' +
+          '  [progressIndicator]="myProgressIndicator"\n' +
+          '>\n' +
+          '</sky-progress-indicator-nav-button>'
+        );
+      }
+
+      this.progressIndicator = this.parentComponent;
+    }
+
+    if (this.buttonType === 'finish') {
+      // The `hasFinishButton` field was added to support legacy API.
+      // Some implementations only include a next button; we cannot
+      // assume that every implementation includes both a finish button and a next button.
+      this.progressIndicator.hasFinishButton = true;
+    }
+
     this.progressIndicator.progressChanges
-    .takeUntil(this.idle)
-    .subscribe((changes: SkyProgressIndicatorChange) => {
-      this.activeIndex = changes.activeIndex;
-      this.changeDetector.detectChanges();
-    });
+      .distinctUntilChanged()
+      .takeUntil(this.ngUnsubscribe)
+      .subscribe((change: SkyProgressIndicatorChange) => {
```

Comment:
Good question. Let me see if I can rework the logic to fit that pattern.

Reply:
More of a discussion than a critique - I notice we do a few things in the logic here that are instigated by changes in the parent progress indicator component. Just curious - why did we decide to have that logic here instead of controlling these types of things in the overall progress indicator component?

In other words, should the nav button components need to know about the world around them? Or should they simply be "dumb" components that house the buttons?

Reply:
More info on this: The nav button is a bit odd because it needs the ability to live both _inside_ the `sky-progress-indicator` component, and _outside_ of it (for use within modals, specifically). Because of this requirement, it needs to inject the parent progress indicator via an `Input`, and watch for progress changes in order to alter its own behavior.

This is how it currently works, unfortunately:
https://github.com/blackbaud/skyux-progress-indicator/blob/master/src/app/public/modules/progress-indicator/progress-indicator-nav-button/progress-indicator-nav-button.component.ts#L32-L33

I tried a few iterations where the nav buttons where "dumber" and didn't inject the progress indicator, but it would require breaking changes. I definitely agree with your observation and think it needs to be reworked in a future major release.

Reply:
Fair enough. I didn't think about the scenario where it had to live both inside and out. Let's move forward with this implementation. I think its a decent compromise!